pkg/http/retry.go

```diff
@@ -0,0 +1,66 @@
+package http
+
+import (
+	"bytes"
+	"io"
+	"math"
+	"net/http"
+	"time"
+)
+
+type RetryableTransport struct {
+	Transport  http.RoundTripper
+	RetryCount int
+}
+
+func (t *RetryableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Clone the request body
+	var bodyBytes []byte
+	if req.Body != nil {
+		bodyBytes, _ = io.ReadAll(req.Body)
+		req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+	}
+	// Send the request
+	resp, err := t.Transport.RoundTrip(req)
+	// Retry logic
+	retries := 0
+	for shouldRetry(err, resp) && retries < t.RetryCount {
+		// Wait for the specified backoff period
+		time.Sleep(backoff(retries))
+		// We're going to retry, consume any response to reuse the connection.
+		drainBody(resp)
+		// Clone the request body again
+		if req.Body != nil {
+			req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+		}
+		// Retry the request
+		resp, err = t.Transport.RoundTrip(req)
+		retries++
+	}
+	// Return the response
+	return resp, err
+}
```

Comment:
we can simplify this

```suggestion
func (t *RetryableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	var bodyBytes []byte
	bodyBytes, err := io.ReadAll(req.Body)
	if err != nil {
		return nil, fmt.Errorf("error reading body: %w", err)
	}

	var resp *http.Response
	retries := -1
	for shouldRetry(err, resp) && retries < t.RetryCount {
		if retries > -1 {
			time.Sleep(backoff(retries))
			// consume any response to reuse the connection.
			drainBody(resp)
		}

		req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
		resp, err = t.Transport.RoundTrip(req)

		retries++
	}

	return resp, err
}
```

Reply:
+1 also lets not ignore any error